bundles/org.openhab.binding.lcn/src/main/java/org/openhab/binding/lcn/internal/LcnModuleDiscoveryService.java

```diff
@@ -0,0 +1,246 @@
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerService;
+import org.openhab.binding.lcn.internal.common.LcnAddrMod;
+import org.openhab.binding.lcn.internal.common.NullScheduledFuture;
+import org.openhab.binding.lcn.internal.connection.Connection;
+import org.openhab.binding.lcn.internal.subhandler.LcnModuleMetaAckSubHandler;
+import org.openhab.binding.lcn.internal.subhandler.LcnModuleMetaFirmwareSubHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Scans all LCN segments for LCN modules.
+ *
+ * Scan approach:
+ * 1. Send "Leerkomando" to the broadcast address with request for Ack set
+ * 2. For every received Ack, send the following requests to the module:
+ * - serial number request (SN)
+ * - module's name first part request (NM1)
+ * - module's name second part request (NM2)
+ * 3. When all three messages have been received, fire thingDiscovered()
+ *
+ * @author Fabian Wolter - Initial Contribution
+ */
+@NonNullByDefault
+public class LcnModuleDiscoveryService extends AbstractDiscoveryService
+        implements DiscoveryService, ThingHandlerService {
+    private final Logger logger = LoggerFactory.getLogger(LcnModuleDiscoveryService.class);
+    private static final Pattern NAME_PATTERN = Pattern
+            .compile("=M(?<segId>\\d{3})(?<modId>\\d{3}).N(?<part>[1-2]{1})(?<name>.*)");
+    private static final int MODULE_NAME_PART_COUNT = 2;
+    private static final int DISCOVERY_TIMEOUT_SEC = 90;
+    private static final int ACK_TIMEOUT_MS = 1000;
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections
+            .unmodifiableSet(Stream.of(LcnBindingConstants.THING_TYPE_MODULE).collect(Collectors.toSet()));
+    private @Nullable PckGatewayHandler bridgeHandler;
+    private Map<LcnAddrMod, Map<Integer, String>> moduleNames = Collections.synchronizedMap(new HashMap<>());
+    private Map<LcnAddrMod, DiscoveryResultBuilder> discoveryResultBuilders = Collections
+            .synchronizedMap(new HashMap<>());
+    private List<LcnAddrMod> successfullyDiscovered = new LinkedList<>();
+    private Queue<@Nullable LcnAddrMod> serialNumberRequestQueue = new ConcurrentLinkedQueue<>();
+    private Queue<@Nullable LcnAddrMod> moduleNameRequestQueue = new ConcurrentLinkedQueue<>();
+    private volatile ScheduledFuture<?> queueProcessor = NullScheduledFuture.getInstance();
```

Comment:
I'm pretty sure the purpose is to allow the fields to be initialized to a non-null value so that he doesn't have to label it as `@Nullable`. Doing so allows him to not have to null check the field ever. I kinda like the solution but you could get the same result by just making the `ScheduledFuture` fields `Future` instead and then initialize them with a completed `CompletableFuture`.